### PR TITLE
Add MemoryManager-option to Storage-struct

### PR DESCRIPTION
**Pull request by przydatek**
Adding an option of using `MemoryManager` in `Storage`-struct.   This enables multiple stable memories for better separation.  
This PR introduces the manager, but does not use it in production -- the actual migration to managed memories will come in a separate PR.

### CODE REVIEW

File: src/internet_identity/src/storage.rs [hunk 1]
```diff
@@ -106,16 +113,137 @@ const STABLE_MEMORY_RESERVE: u64 = 8 * GB / 10;
 
 const PERSISTENT_STATE_MAGIC: [u8; 4] = *b"IIPS"; // II Persistent State
 
+/// MemoryManager parameters.
+const ANCHOR_MEMORY_ID: MemoryId = MemoryId::new(0);
+const BUCKET_SIZE_IN_PAGES: u16 = 1024;
```
Comment (frederikrothenberger) [inline src/internet_identity/src/storage.rs:118]: Why did you chose this bucket size? Maybe add a comment here to explain.

Also, 1024 is quite large, the (new) default is 128. We might end up wasting quite some space, if we allocate `VirtualMemory`s for smallish things (e.g. asset storage, fixed persistent state, etc.).

**Review by frederikrothenberger — COMMENTED** [review]
Looks good!
The only important point is that we should have a discussion on the MemoryManager bucket size.

File: src/internet_identity/src/storage.rs [hunk 2]
```diff
@@ -106,16 +113,137 @@ const STABLE_MEMORY_RESERVE: u64 = 8 * GB / 10;
...
 
+enum AnchorMemory<M: Memory> {
+    Single(RestrictedMemory<M>),
+    Managed(VirtualMemory<RestrictedMemory<M>>),
+}
+
+// Auxiliary traits and structures to encapsulate read/write operations
+// to different flavours of anchor memory.
+trait MemoryWriter {
+    fn write(&mut self, buf: &[u8]) -> Result<usize, Error>;
```
[the remark this answers is not in the document]
Comment (przydatek) [reply src/internet_identity/src/storage.rs]: Good point, removed.

File: src/internet_identity/tests/integration/stable_memory.rs [hunk 3]
```diff
@@ -279,6 +279,65 @@ fn should_read_persistent_state() -> Result<(), CallError> {
...
 }
 
+#[allow(dead_code)]
+fn prepare_persistent_state_v7() {
+    let env = env();
+    let canister_id =
+        install_ii_canister_with_arg(&env, II_WASM.clone(), arg_with_anchor_range((127, 129)));
+
+    let challenge = api::create_challenge(&env, canister_id).expect("challenge creation failed");
+    let register_response = api::register(
```
[the remark this answers is not in the document]
Comment (przydatek) [reply src/internet_identity/tests/integration/stable_memory.rs]: Thanks, done!

File: src/internet_identity/tests/integration/stable_memory.rs [hunk 4]
```diff
@@ -279,6 +279,65 @@ fn should_read_persistent_state() -> Result<(), CallError> {
...
+    save_compressed_stable_memory(
+        &env,
+        canister_id,
+        "stable_memory/persistent_state_no_archive_v7.bin.gz",
+        "persistent_state_no_archive_v7.bin",
+    )
+}
+
+#[test]
+fn should_read_persistent_state_v7() -> Result<(), CallError> {
```
[the remark this answers is not in the document]
Comment (przydatek) [reply src/internet_identity/tests/integration/stable_memory.rs:287]: I suppose the comment refers to `prepare_persistent_state_v7` above, not to this test (which is not dead code). I moved to a separate module.

File: src/internet_identity/src/storage.rs (hunk 1 above)
Comment (przydatek) [reply src/internet_identity/src/storage.rs:118]: Yep, this is up for discussion -- let's chat and I'll change it afterwards.

**Review by przydatek — COMMENTED** [review]
Thanks for the review, let's discuss the bucket size in person.

File: src/internet_identity/src/storage.rs (hunk 1 above)
Comment (przydatek) [reply src/internet_identity/src/storage.rs]: As discussed, I changed the bucket size to 128 (and added a comment).

**Review by frederikrothenberger — APPROVED** [review]
LGTM!

**Pull request merged**